## The Goal
- #### Learn some common programming patterns 
- #### Improve algorithmic skills
- #### Improve job prospects






## Notebook Structure
- #### Whirlwind tour of list, str, set, tuple and dict
- #### First look at Big O
- #### Search and Sort introduction => Linear, Binary, HashTable, "sorted" vs "sort"
- #### Array manipulation => Traversal(linear, both-ends), rotation, duplicate/palindromic check, reverse words

### list: Can hold anything in order, best friend
### tuple: List but immutable(Change is forbidden)
### str: string, immutable
### set: list without duplicates
### dict: dictionary, key-value pairs, second-best friend

In [463]:
a = [1,2,3,'list', [4]]

In [464]:
len(a)

5

In [465]:
print(a)
print(a[:2])
print(a[0:2])
print(a[2:])
print(a[1:3])
print(a[0:4:2])
print(a[::-1]) #reverse

[1, 2, 3, 'list', [4]]
[1, 2]
[1, 2]
[3, 'list', [4]]
[2, 3]
[1, 3]
[[4], 'list', 3, 2, 1]


In [466]:
print(a[2])
a[2] = 100
print(a[2])

3
100


In [467]:
a

[1, 2, 100, 'list', [4]]

In [468]:
type(a[3])

str

In [469]:
mystring = "demo"
mystring

'demo'

In [470]:
mystring = "a bigger demo"
mystring

'a bigger demo'

In [471]:
mystring[2:9]

'bigger '

In [472]:
print(mystring[3])
mystring[3] = 'e'

i


TypeError: 'str' object does not support item assignment

In [473]:
strlist = list(mystring)
strlist

['a', ' ', 'b', 'i', 'g', 'g', 'e', 'r', ' ', 'd', 'e', 'm', 'o']

In [474]:
print(strlist[3])
strlist[3] = 'e'
strlist[3]

i


'e'

In [475]:
strlist

['a', ' ', 'b', 'e', 'g', 'g', 'e', 'r', ' ', 'd', 'e', 'm', 'o']

In [476]:
newstr = ''.join(strlist)  # common idiom to change from list to str

In [477]:
newstr

'a begger demo'

In [478]:
strlist.append(123)

In [479]:
topple = tuple(strlist)

In [480]:
len(topple)

14

In [481]:
topple.append(3)

AttributeError: 'tuple' object has no attribute 'append'

In [482]:
topple[4]

'g'

In [483]:
topple[4] = 3

TypeError: 'tuple' object does not support item assignment

In [484]:
wonderset = set(strlist)

In [485]:
len(wonderset)

10

In [486]:
wonderset

{' ', 123, 'a', 'b', 'd', 'e', 'g', 'm', 'o', 'r'}

In [487]:
wonderset.append(123)

AttributeError: 'set' object has no attribute 'append'

In [488]:
wonderset.add(123)
wonderset.add('12')

In [489]:
wonderset  # As we can see, the order has been changed.

{' ', '12', 123, 'a', 'b', 'd', 'e', 'g', 'm', 'o', 'r'}

In [490]:
d = {}

In [491]:
d[1] = 1
d[2] = 1
d[3] = 2
d[4] = 3
d[5] = 5

In [492]:
d

{1: 1, 2: 1, 3: 2, 4: 3, 5: 5}

In [493]:
for i in d.keys():
    print(d[i])

1
1
2
3
5


In [494]:
for k, v in d.items():
    print(k, v)

1 1
2 1
3 2
4 3
5 5


In [495]:
letters = {}
letters['a'] = 34
letters['b'] = 1
letters[3] = 2

In [496]:
for i, v in enumerate(letters):
    print(f'{i}:{v}')

0:a
1:b
2:3






## Big O

####  Mathematical way to assess performance: For large inputs, how long will this program take to run , *in the worst case*?

##### Always think: Can we do better? 

### Thumb rules: 

- constants don't count
- how many times will the program access the input values?

## Remember:
- O(1) >> O(LogN) >> O(N) >> O(NLogN) >> O(N\**2) >> O(N\**k) >> O(k\**N)
- O(1) => constant access
- O(LogN) => search space divides in half in each operation
- O(N) => need to access every element 
- O(NLogN) => comparison based sorting or combination of N and LogN
- O(N\**k) => each element is accessed k times.
- O(k\**N) => exponential

In [497]:
listi = [1,2,3,4,5,6,7,8,9]
listi

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [498]:
def geteven(listi):
    newlist = []
    for i in listi:
        if i%2==0:
            newlist.append(i)
    return newlist

In [499]:
geteven(listi)

[2, 4, 6, 8]

In [500]:
def getodd(listi):
    newlist = []
    for i in listi:
        if i%2:         
            newlist.append(i)
    return newlist

In [501]:
getodd(listi)

[1, 3, 5, 7, 9]

In [502]:
def doubly(listi):
    newlisti = []
    for i in listi:
        for j in listi:
            newlisti.append(i+j)
    return newlisti

In [503]:
print(doubly(listi))

[2, 3, 4, 5, 6, 7, 8, 9, 10, 3, 4, 5, 6, 7, 8, 9, 10, 11, 4, 5, 6, 7, 8, 9, 10, 11, 12, 5, 6, 7, 8, 9, 10, 11, 12, 13, 6, 7, 8, 9, 10, 11, 12, 13, 14, 7, 8, 9, 10, 11, 12, 13, 14, 15, 8, 9, 10, 11, 12, 13, 14, 15, 16, 9, 10, 11, 12, 13, 14, 15, 16, 17, 10, 11, 12, 13, 14, 15, 16, 17, 18]


## Search + Sort (High Level)

In [504]:
a = [x**2 for x in range(100)]
print(a)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801]


In [505]:
print(53 in a)
print(4096 in a)

False
True


In [506]:
def linear_search(a, k):
    for i in a:
        if i==k:
            return True
    return False

In [507]:
linear_search(a, 53)

False

In [508]:
linear_search(a, 4096)

True

In [509]:
def non_pythonic_linear_search(a, k):
    for i in range(len(a)):
        if a[i] == k:
            return True
    return False

In [510]:
non_pythonic_linear_search(a, 53)

False

In [511]:
non_pythonic_linear_search(a, 4096)

True

In [514]:
# How do you find a word in a dictionary?

# 1-2-3-4-5-6-7-8, 6
# low=0, high=7, mid = 4, a[mid] = 5
# low=5, high=7, mid = 6, a[mid] = 7
# low=5, high=6, mid = 5, a[mid] = 6

def binary_search(a, k):
    
    if(k is None or a is None or not a):  # pythonic idioms
        return None
    
    high = len(a)-1  #because we need the range
    low = 0
    while(low<=high):
        mid = low + (high-low)//2  # to avoid integer overflow
        if(k == a[mid]):
            return mid
        
        elif(k > a[mid]):
            low = mid+1

        else:
            high = mid-1
            
    return None
    

In [515]:
empty_a=[]
if not empty_a :
    print(2) 

2


In [516]:
# Tests for binary search

assert binary_search(a, 53) == None
assert binary_search(a, 4096) == 64

assert binary_search(a,-23) == None
assert binary_search(a, 3.4) == None


assert binary_search(empty_a, 4096) == None
assert binary_search(a,None) == None

In [517]:
print(2/3)
print(2//3)

0.6666666666666666
0


In [518]:
2/3

0.6666666666666666

In [519]:
2//3

0

In [520]:
print(binary_search([10, 3, 45, 64, 1, 0, -32], 1))

None


In [521]:
a = [10, 3, 45, 64, 1, 0, -32]

In [522]:
def search(a, k):
    a.sort()  # in-place sorting
    return binary_search(a, k)

In [523]:
a.sort()
a

[-32, 0, 1, 3, 10, 45, 64]

In [524]:
a = [10, 3, 45, 64, 1, 0, -32]

In [525]:
assert search(a, 53) == None
assert search(a, 64) == 6

assert search(a,-32) == 0
assert search(a, 3.4) == None


assert search(empty_a, 4096) == None
assert search(a,None) == None

In [526]:
sorted_list = sorted(a)
binary_search(sorted_list, 10)

4

Binary Search takes O(logN) time in worst case, as our search-space gets halved everytime. Can, we do better?

### Introducing, HashTables!! 

- for all practical purposes, we can use a dict.
- takes O(N) space
- uses hashing and array-access for looking up items

In [527]:
ht = {}
for i in a:
    ht[i]= True
ht

{-32: True, 0: True, 1: True, 3: True, 10: True, 45: True, 64: True}

In [528]:
-32 in ht

True

In [529]:
90 in ht

False

### How is everything stored in memory?


Imagine this to be a snapshot of a memory location and its surroundings: 121-122-123-124....

- A pointer to memory location 121 would be an object whose value would be the address 121
- An array has contiguous memory. So, we could access the 4th element in an array by adding 4 to the memory address of the 0th element
- the "4" in reality is the size of the type we are trying to store
- From the Python FAQ, for lists, "the implementation uses a contiguous array of references to other objects, and keeps a pointer to this array and the array’s length in a list head structure." 

Thus, arrays and in turn lists can give us O(1) access time for any data. This property is used in dict.

### Logical model of Python Hash table 

Taken from : [here](https://stackoverflow.com/questions/327311/how-are-pythons-built-in-dictionaries-implemented)
```-+-----------------+
0| <hash|key|value>|
-+-----------------+
1|      ...        |
-+-----------------+
.|      ...        |
-+-----------------+
i|      ...        |
-+-----------------+
.|      ...        |
-+-----------------+
n|      ...        |
-+-----------------+```


In [530]:
id(a)

4400475344

In [531]:
a = 2
b = 2

In [532]:
a==b

True

In [533]:
a is b

True

In [534]:
id(a)==id(b)

True

In [535]:
a = []
b = []
a is b

False

In [536]:
print(f'Id of a : {id(a)}')
print(f'Id of b : {id(b)}')

Id of a : 4400020368
Id of b : 4400019888


In [537]:
a is b 

False

In [538]:
a = [1,2]

def add(b):
    b.append(3)

add(a)
a

[1, 2, 3]

In [543]:
b = "i am inevitable"
def add(b):
    b = "i am iron man"
add(b)
b

'i am inevitable'

In [544]:
hash('565')

-5739472448459660605

In [545]:
hash('qwerty')%256

165

In [549]:
class Hashmap():
    size = 34
    def __init__(self):
        self.values = [0]*self.size
        self.keys = set()
    def findHash(self, k):
        return hash(k)%self.size
    def add(self, k, v):
        self.values[self.findHash(k)]=v
        self.keys.add(k)
    def get(self, k):
        return self.values[self.findHash(k)]
        

In [559]:
hmap = Hashmap()
hmap.add(1, "23")
hmap.add(34, 45)
hmap.add('12', 'qwerty')
print(hmap.findHash(34)) # returns 0
print(hmap.values[0])

0
45


In [560]:
hmap.get(1)

'23'

In [561]:
hmap.get(34)

45

In [562]:
hmap.get('12')

'qwerty'

### Collision in Hashing:

- Chaining (list at list)


```0| <hash|key|value>|
-+-----------------+
1|      ...        |
-+-----------------+
.|      ...        |
-+-----------------+-----------------+-----------------+
i|    occupied     |  place here     |  place here     |
-+-----------------+-----------------+-----------------+
.|      ...        |
-+-----------------+
n|      ...        |
-+-----------------+```

- Open Addressing with Probing (use nearby memory location)

```0| <hash|key|value>|
-+-----------------+
1|      ...        |
-+-----------------+
.|      ...        |
-+-----------------+
k|     occupied    |                   
-+-----------------+
.|     place here  |
-+-----------------+
n|      ...        |
-+-----------------+```

```





```

## List based Problems


In [563]:
# cyclically rotate a string to its left by k elements: 12345, 2 => 34512
# Algorithm: 12345 => 543 21 => 345 12=>34512

def mutate_string(string,k):
    s = list(string)
    
    s = s[::-1]
    s = s[k::-1] + s[:k:-1]
   
    return ''.join(s)

In [564]:
a =[1,2,3,4,5, 6, 7]
a[2::-1]+a[:2:-1] 

[3, 2, 1, 7, 6, 5, 4]

In [565]:
a = "1234567"

In [566]:
mutate_string(a, 2)

'5671234'

In [567]:
# find duplicates
def find_dup(a):
    nodups = dict()
    for i in a:
        if i in nodups:
            return i
        else:
            nodups[i] = True
    return None

In [568]:
find_dup([1, 34, 5, 1])

1

In [569]:
print(find_dup([1,2, 3]))

None


In [570]:
# find two numbers which sum to X
# 2, 6, 3, 4, 2, 5, 1 => 7 : 3,4  # Think how can we get 4 if we have 3. 

def find_summer(a, x):
    if not a :
        return
    ht = dict()
    
    for i in a:
        if (x-i) in ht:
            return x-i, i
        else:
            ht[i] = x-i
    return None

In [571]:
a = [2, 6, 3, 4, 2, 5, 1]
print(find_summer([],2))
print(find_summer(a,5))
print(find_summer(a,10))

None
(2, 3)
(6, 4)


In [572]:
# check palindrome
# 12321 => True 
# 123431 => False

def is_palindrome(a):
    if not a:
        return
    a = list(a)
    return a[::-1] == a

In [573]:
is_palindrome([1,2,1])

True

In [574]:
is_palindrome("abcba")

True

In [575]:
is_palindrome("abcagh")

False

In [576]:
def is_palindrome(a):
    size = len(a)
    i = 0
    j = size-1
    
    while(i<=j):
        if a[i]!=a[j] :
            return False
        i+=1
        j-=1
    return True

In [577]:
is_palindrome([1,2,1])

True

In [578]:
is_palindrome("abcba")

True

In [579]:
is_palindrome("abca")

False

In [580]:
# check if anagrams
# cat=>tca=>tac..

def is_anagram(a,b):
    return sorted(list(a))==sorted(list(b))

In [581]:
is_anagram("cat", "tca")

True

In [582]:
## what does it mean to be an anagram?? 

def is_anagram(a,b):
    if(len(a)!=len(b)):
        return False
    
    ht = dict()
    
    for i in a:
        if i in ht:
            ht[i] += 1
        else:
            ht[i] = 1
    
    for j in b:
        if j not in ht:
            return False
        else:
            ht[j] -= 1
    
    for k in ht:
        if ht[k] != 0:
            return False
    
    return True
        

In [583]:
is_anagram("cattac", "tcaactkjeekjje")

False

In [584]:
is_anagram("cattac", "tcacct")

False

In [585]:
# Reverse words in a string: esreveR sdrow ni a gnirts

def reverse_words(string):
    s = list(string)
    ret = []    
    size = len(s)
    start = 0
    
    for i in range(size):
        if s[i] == " ":
            ret += reverse(s[start:i])
            ret += " "
            start = i+1
    ret += reverse(s[start:size]) # IMPORTANT : To handle trailing case
    return ''.join(ret)
        
def reverse(a):
    size = len(a)
    i = 0 
    j = size-1
    while(i<=j):
        a[i], a[j] = a[j], a[i]  #Pythonic , otherwise a 3rd variable or arithmetical/logical operators may be needed 
        i+=1
        j-=1
    return a       

In [586]:
reverse_words("asdf")

'fdsa'

In [587]:
reverse_words("Reverse words in a string")

'esreveR sdrow ni a gnirts'

In [588]:
a

[2, 6, 3, 4, 2, 5, 1]

In [589]:
reverse(a)

[1, 5, 2, 4, 3, 6, 2]

In [590]:
# Pythonic way
def reverse_words(a):
    words = a.split(" ") 

    newWords = [word[::-1] for word in words] 
      
    return " ".join(newWords) 

In [591]:
def swap(a,b):
    temp = a
    a = b
    b = temp
    return a,b
    
def superswap(a, b):
    a = a+b
    b = a-b
    a = a-b
    return a,b

In [592]:
a = 1
b = 2
print(swap(a,b))
print(superswap(a,b))

(2, 1)
(2, 1)


```




```

## Practice problems:

 #### Change Binary Search algorithm to solve this problem 
 - Given a sorted array that can contain duplicates, find the first occurrence of the
target element. For example:
A = [1, 3, 14, 23, 23, 23, 56] and Target = 23, return index 3. 

####  Find two numbers that sum to X if the array is sorted. 
- For example: [1,2,3,4,5,6,7] => 7 : 3,4
- Hint: Try traversing the list from both ends ;)


Next time, we'll delve into 
- #### More uses for Binary Search
- #### Matrix manipulation => rotation, matrix spiral 
- #### More Array/List manipulation
- #### Implementing Sorting